File: go/loyalty/activities.go
package loyalty

import (
	"context"
	"errors"
	"go.temporal.io/sdk/temporal"

	"go.temporal.io/api/enums/v1"
	"go.temporal.io/api/serviceerror"
	"go.temporal.io/sdk/activity"
	"go.temporal.io/sdk/client"
)

type Activities struct {
	Client client.Client
}

func (*Activities) SendEmail(ctx context.Context, body string) error {
	logger := activity.GetLogger(ctx)
	logger.Info("Sending email.", "Contents", body)
	return nil
}

func (a *Activities) StartGuestWorkflow(ctx context.Context, guest CustomerInfo) error {
	logger := activity.GetLogger(ctx)

	workflowOptions := client.StartWorkflowOptions{
		TaskQueue:             TaskQueue,
		WorkflowIDReusePolicy: enums.WORKFLOW_ID_REUSE_POLICY_REJECT_DUPLICATE,
	}

	logger.Info("Starting and signaling guest workflow.", "GuestID", guest.CustomerID)
	_, err := a.Client.SignalWithStartWorkflow(ctx, CustomerWorkflowID(guest.CustomerID),
		SignalEnsureMinimumStatus, StatusLevelForPoints(guest.LoyaltyPoints).Ordinal,
		workflowOptions, CustomerLoyaltyWorkflow, guest, true)

	target := &serviceerror.WorkflowExecutionAlreadyStarted{}
	if errors.As(err, &target) {
		return temporal.NewApplicationError("Guest account cannot be recreated from a closed status.",
			"GuestAlreadyCanceledError")
	}
	return nil
}

File: python/activities.py
from temporalio import activity, common
from temporalio.client import Client
from temporalio.exceptions import WorkflowAlreadyStartedError

from shared import Customer


class LoyaltyActivities:
    def __init__(self, client: Client) -> None:
        self.client: Client = client

    @activity.defn
    async def send_email(self, body: str) -> None:
        activity.logger.info("Sending email with contents %s" % body)

    @activity.defn
    async def start_guest_workflow(self, guest: Customer) -> bool:
        from loyalty_workflow import (
            CustomerLoyaltyWorkflow,
            TASK_QUEUE,
            SIGNAL_ENSURE_MINIMUM_STATUS
        )

        activity.logger.info("Starting guest workflow with ID %s" % guest.id)

        try:
            await self.client.start_workflow(
                CustomerLoyaltyWorkflow.run,
                id=CustomerLoyaltyWorkflow.workflow_id(guest.id),
                task_queue=TASK_QUEUE,
                start_signal=SIGNAL_ENSURE_MINIMUM_STATUS,
                start_signal_args=[guest.tier],
                id_reuse_policy=common.WorkflowIDReusePolicy.ALLOW_DUPLICATE_FAILED_ONLY,
                args=[guest, True],
            )
        except WorkflowAlreadyStartedError:
            return False

        return True
